GovernedLotteryHook.sol

```diff
@@ -0,0 +1,188 @@
+// SPDX-License-Identifier: GPL-3.0-or-later
+
+pragma solidity ^0.8.24;
+
+import { IERC20 } from "@openzeppelin/contracts/token/ERC20/IERC20.sol";
+import { Ownable } from "@openzeppelin/contracts/access/Ownable.sol";
+import { SafeERC20 } from "@openzeppelin/contracts/token/ERC20/utils/SafeERC20.sol";
+
+import { IHooks } from "@balancer-labs/v3-interfaces/contracts/vault/IHooks.sol";
+import { IRouterCommon } from "@balancer-labs/v3-interfaces/contracts/vault/IRouterCommon.sol";
+import { IVault } from "@balancer-labs/v3-interfaces/contracts/vault/IVault.sol";
+import {
+    AfterSwapParams,
+    LiquidityManagement,
+    SwapKind,
+    TokenConfig,
+    HookFlags
+} from "@balancer-labs/v3-interfaces/contracts/vault/VaultTypes.sol";
+
+import { EnumerableMap } from "@balancer-labs/v3-solidity-utils/contracts/openzeppelin/EnumerableMap.sol";
+import { FixedPoint } from "@balancer-labs/v3-solidity-utils/contracts/math/FixedPoint.sol";
+import { VaultGuard } from "@balancer-labs/v3-vault/contracts/VaultGuard.sol";
+import { BaseHooks } from "@balancer-labs/v3-vault/contracts/BaseHooks.sol";
+
+contract GovernedLotteryHook is BaseHooks, VaultGuard, Ownable {
+    using FixedPoint for uint256;
+    using EnumerableMap for EnumerableMap.IERC20ToUint256Map;
+    using SafeERC20 for IERC20;
+
+    // Governance Proposal Struct
+    struct Proposal {
+        uint256 proposalId;
+        string description;
+        uint64 newSwapFeePercentage;
+        uint8 newLuckyNumber;
+        uint256 votesFor;
+        uint256 votesAgainst;
+        uint256 votingDeadline;
+    }
+
+    // State variables for proposals
+    Proposal[] public proposals;
+    mapping(uint256 => mapping(address => bool)) public hasVoted;
+
+    // Lottery and fee variables
+    uint8 public LUCKY_NUMBER = 10;
+    uint8 public constant MAX_NUMBER = 20;
+    uint64 public hookSwapFeePercentage;
+
+    EnumerableMap.IERC20ToUint256Map private _tokensWithAccruedFees;
+    uint256 private _counter = 0;
+    address private immutable _trustedRouter;
+
+    // Events for governance and fees
+    event ProposalCreated(uint256 proposalId, string description);
+    event VoteCast(uint256 proposalId, address voter, bool support);
+    event ProposalImplemented(uint256 proposalId, uint64 newSwapFeePercentage, uint8 newLuckyNumber);
+    event LotteryWinningsPaid(
+        address indexed hooksContract,
+        address indexed winner,
+        IERC20 indexed token,
+        uint256 amountWon
+    );
+
+    constructor(IVault vault, address router) VaultGuard(vault) Ownable(msg.sender) {
+        _trustedRouter = router;
+    }
+
+    // Create a new governance proposal
+    function createProposal(
```

Comment:
So in this design, the owner can propose multiple changes, then choose which one(s) to implement. With 5 proposals, if 3 were approved, the owner could alternate between all the approved ones.

If only the owner can do this, consider one proposal at a time (perhaps with a variable deadline, constrained by a hard-coded minimum; e.g., min 3 days, but they could pick 3, 5, 7, 14 days, etc.), and a permissionless `implementProposal`. Only the owner could propose, but anyone could enforce the result after the deadline. Also consider an "implemented" flag, so that `implementProposal` can only be called once. Of course, at that point, since the owner has total control anyway, you might as well just let the owner set the fee directly... which I see you're also doing, so the voting mechanism doesn't really do anything.

Another possible design (though more complex), would be some notion of epochs: time periods during which anyone can create a proposal (probably capped to a certain number, like 5, especially if you're using an iterable array), and a permissionless "implement" that would implement the highest scoring proposal (or do nothing if there were no votes), then clear the data and start a new epoch.

Pending proposals could only be queried off-chain, which is maybe ok.